File: langchain/appendix_A/basic.py
import uuid
from pathlib import Path
from dotenv import load_dotenv

# Load environment variables
env_path = Path(__file__).resolve().parent / "../../shared/.env"
load_dotenv(dotenv_path=env_path)

# LangChain and LangGraph imports
from langchain.tools import tool
from langchain_core.messages import HumanMessage, SystemMessage
from langchain_openai import ChatOpenAI
from langgraph.graph import START, END, MessagesState, StateGraph
from langgraph.prebuilt import ToolNode, tools_condition


# Define Tool
@tool
def generate_uuid_tool():
    """Generate a unique UUID identifier."""
    return {
        "uuid": str(uuid.uuid4())
    }

tools = [generate_uuid_tool]
llm = ChatOpenAI(model="gpt-5.6-luna", reasoning_effort="none").bind_tools(tools)


# Node function to unpack state["messages"] for the LLM
def call_model(state: MessagesState):
    response = llm.invoke(state["messages"])
    return {"messages": [response]}


# Construct Graph
builder = StateGraph(MessagesState)

builder.add_node("model", call_model)
builder.add_node("tools", ToolNode(tools))

builder.add_edge(START, "model")
builder.add_conditional_edges("model", tools_condition)
builder.add_edge("tools", "model")
builder.add_edge("model", END)


graph = builder.compile()


# Invoke Graph
result = graph.invoke(
    {
        "messages": [
            SystemMessage(
                content="Use generate_uuid when user asks for UUID. Keep responses short."
            ),
            HumanMessage(
                content="Generate a ticket ID for last night's network outage"
            ),
        ]
    }
)

print(result["messages"][-1].content)

File: langchain/appendix_A/basic.ts
import path from "path";
import { fileURLToPath } from "url";
import dotenv from "dotenv";

// Load environment variables from relative .env file
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
dotenv.config({ path: path.resolve(__dirname, "../../shared/.env") });

// LangChain and LangGraph imports
import { tool } from "@langchain/core/tools";
import { HumanMessage, SystemMessage } from "@langchain/core/messages";
import { ChatOpenAI } from "@langchain/openai";
import { StateGraph, MessagesAnnotation, START, END } from "@langchain/langgraph";
import { ToolNode, toolsCondition } from "@langchain/langgraph/prebuilt";
import { randomUUID } from "node:crypto";
import { z } from "zod";

// Define Tool (uses Zod for schema validation)
const generateUUID = tool(
    async () => randomUUID(),
    {
      name: "generate_uuid",
      description: "Generate a unique UUID",
      schema: z.object({}), // no arguments  
    }
);

const tools = [generateUUID];
const llm = new ChatOpenAI({
    model: "gpt-5.6-luna",
    modelKwargs: { reasoning_effort: "none"}
}).bindTools(tools);

// Node function to unpack state.messages for the LLM
async function callModel(state: typeof MessagesAnnotation.State) {
  const response = await llm.invoke(state.messages);
  return { messages: [response] };
}

// Construct Graph 
const builder = new StateGraph(MessagesAnnotation)
  .addNode("model", callModel)
  .addNode("tools", new ToolNode(tools))
  .addEdge(START, "model")
  .addConditionalEdges("model", toolsCondition)
  .addEdge("tools", "model")
  .addEdge("model", END);

const graph = builder.compile();

// Invoke Graph
const result = await graph.invoke({
  messages: [
    new SystemMessage(
      "Use generate_uuid when user asks for UUID. Keep responses short."
    ),
    new HumanMessage("Generate a ticket ID for last night's network outage"),
  ],
});

// Output final response
const lastMessage = result.messages[result.messages.length - 1];
console.log(lastMessage.content);
